common/flatpak-dir.c

```diff
@@ -4018,19 +4022,231 @@ repo_get_remote_collection_id (OstreeRepo *repo,
+          return NULL;
+        }
+
+      token_cache_filename = g_strconcat (g_file_get_path (cache_dir),
+                                          "/api-server-token-",
+                                          token_url_checksum,
+                                          NULL);
+      token_cache_file = g_file_new_for_path (token_cache_filename);
+      token_cache_file_info = g_file_query_info (token_cache_file,
+                                                 G_FILE_ATTRIBUTE_TIME_MODIFIED,
+                                                 G_FILE_QUERY_INFO_NONE,
+                                                 cancellable, NULL);
+      g_file_info_get_modification_time (token_cache_file_info, &token_cache_mtime);
+      g_get_current_time (&current_time);
+
+      /* Tokens expires in one hour, we won't reuse tokens that are close
+       * to expire, that's why we check for at most 3400 seconds old tokens */
+      if (g_file_query_exists (token_cache_file, cancellable) &&
+          current_time.tv_sec - token_cache_mtime.tv_sec < 3400)
+        {
+          if (!g_file_load_contents (token_cache_file,
+                                     cancellable,
+                                     &token_buffer,
+                                     NULL, NULL,
+                                     error))
+            {
+              g_warning ("Unable to read token cache file %s: %s",
+                         token_cache_filename,
+                         (*error)->message);
+              g_clear_pointer (&token_buffer, g_free);
+            }
+        }
+
+      if (token_buffer)
+        return token_buffer;
+
+      ensure_soup_session (dir);
+      request = soup_session_request_http (dir->soup_session, "GET",
+                                           token_url, error);
+      if (!request)
+        {
+          g_prefix_error (error, "Unable to create repo token Soup request: ");
+          return NULL;
+        }
+
+      http_message = soup_request_http_get_message (request);
+      bearer_header = g_strconcat ("Bearer: ", access_token, NULL);
+      soup_message_headers_replace (http_message->request_headers, "Authorization", bearer_header);
+
+      context = flatpak_main_context_new_default ();
+      soup_request_send_async (SOUP_REQUEST (request), cancellable,
+                               async_result_cb, &request_send_result);
+
+      while (request_send_result == NULL)
+        g_main_context_iteration (context, TRUE);
+
+      request_stream = soup_request_send_finish (SOUP_REQUEST (request),
+                                                 request_send_result,
+                                                 error);
+      if (!request_stream)
+        {
+          g_prefix_error (error, "Unable to create repo token Soup request: ");
+          return NULL;
+        }
+
+      if (!SOUP_STATUS_IS_SUCCESSFUL (http_message->status_code))
+        {
+          *error = g_error_new (G_IO_ERROR, G_IO_ERROR_FAILED,
```

Comment:
You can’t assume that `error` will be non-`NULL`. Use `g_set_error (error, …)` to set the error.

See the [summary documentation on using `GError`](https://developer.gnome.org/glib/stable/glib-Error-Reporting.html#gerror-rules).